exec/exec.go

```diff
@@ -14,33 +15,65 @@ import (
 
 var ErrWaitPIDTimeout = fmt.Errorf("Timed out waiting for PID to complete")
 
+type CmdOpts struct {
+	timeout time.Duration
+}
+
 // RunCommandExt is a convenience function to run/log a command and return/log stderr in an error upon
-// failure. Chops off a trailing newline (if present)
-func RunCommandExt(cmd *exec.Cmd) (string, error) {
-	cmdStr := strings.Join(cmd.Args, " ")
-	log.Info(cmdStr)
-	outBytes, err := cmd.Output()
+// failure.
+func RunCommandExt(cmd *exec.Cmd, opts CmdOpts) (string, error) {
+
+	// log in a way we can copy-and-paste into a terminal
+	args := strings.Join(cmd.Args, " ")
+	log.WithFields(log.Fields{"dir": cmd.Dir}).Info(args)
+
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+
+	start := time.Now()
+	err := cmd.Start()
 	if err != nil {
-		exErr, ok := err.(*exec.ExitError)
-		if !ok {
-			return "", errors.WithStack(err)
-		}
-		errOutput := string(exErr.Stderr)
-		log.Errorf("`%s` failed: %s", cmdStr, errOutput)
-		return "", errors.New(strings.TrimSpace(errOutput))
+		return "", err
+	}
+
+	done := make(chan error)
+	go func() { done <- cmd.Wait() }()
+
+	// Start a timer
+	timeout := 300 * time.Second
```

Comment:
I don't see  a way to run without a timeout. Also I feel the default should be no-timeout since 300 seconds is arbitrary. Workflow runs commands that run indefinitely (e.g. `docker wait CONTAINER`). For Argo CD, we will have a shared, global CmdOpts{} that most callers will use, which **will** have a timeout set.